File: example/Tensor/contiguous.py
import sys
from pathlib import Path
home = str(Path.home())
sys.path.append(home + '/Cytnx_lib')
from cytnx import *

A = Tensor([3,4,5])
print(A.shape())

B = A.permute(0,2,1)
print(B.shape())

C = B.contiguous()
print(B.is_contiguous()) #false
print(C.is_contiguous()) #true
print(C.shape())

File: example/Tensor/contiguous.cpp
#include "cytnx.hpp"
#include <iostream>

using namespace cytnx;
int main() {
  Tensor A({3, 4, 5});
  std::cout << A.shape() << std::endl;

  Tensor B = A.permute({0, 2, 1});
  std::cout << B.shape() << std::endl;

  //[Note] permute will not actually move the internal memory (storage) layout.
  //       this is called non-contiguous status.
  //       the memory layout will only move when Tensor.contiguous() is called.
  Tensor C = B.contiguous();  // actual moving the memory
  std::cout << B.is_contiguous() << std::endl;  // false.
  std::cout << C.is_contiguous() << std::endl;  // true.
  std::cout << C.shape() << std::endl;

  std::cout << C.same_data(B) << std::endl;  // false
  std::cout << B.same_data(A) << std::endl;  // true

  return 0;
}
